### PR TITLE
feat: add karpenter-provioners submodule

### PR DESCRIPTION
**Pull request by yhrunyk**

### CODE REVIEW

File: modules/karpenter-provisioners/variables.tf [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+        evictionMaxPodGracePeriod   = 600
+        imageGCHighThresholdPercent = 85
+        imageGCLowThresholdPercent  = 80
+        cpuCFSQuota                 = true
+        podsPerCore                 = 2
+        maxPods                     = 110
+    })
+
+  }))
+  default = [
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:156]: should be an empty list by default

File: modules/karpenter-provisioners/variables.tf [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+        deviceName = "/dev/xvda"
+        ebs = {
+          volumeType          = "gp3"
+          volumeSize          = "35Gi"
+          deleteOnTermination = true
+          encrypted           = true
+        }
+    })
+  }))
+  default = [
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:396]: should be an empty list by default

File: modules/karpenter-provisioners/variables.tf [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+    requirements = optional(
+      list(object(
+        {
+          key      = string
+          operator = string
+          values   = list(string)
+        }
+        )), [
+        {
+          key      = "karpenter.sh/capacity-type"
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:37]: keep only `karpenter.sh/capacity-type` and `kubernetes.io/arch` and remove the rest of things

File: modules/karpenter-provisioners/variables.tf [hunk 4]
```diff
@@ -0,0 +1,445 @@
+variable "karpenter_node_pools" {
+  type = list(object({
+    name                      = string
+    instance_types            = list(string)
+    kubernetes_labels         = map(string)
+    annotations               = optional(map(string), null)
+    ttl_seconds_after_empty   = optional(number, 300)
+    ttl_seconds_until_expired = optional(number, 2592000)
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:8]:
```suggestion
    ttl_seconds_until_expired = optional(number, null)
```

File: modules/karpenter-provisioners/variables.tf [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+    weight                    = optional(number, 0)
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:9]:
```suggestion
    weight                    = optional(number, null)
```

File: modules/karpenter-provisioners/main.tf [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  karpenter_provisioner_values = yamlencode({ provisioners = [for node in var.karpenter_node_pools :
+    {
+      name = node.name
+      spec = {
+        ttlSecondsAfterEmpty = node.ttl_seconds_after_empty
+        labels               = node.kubernetes_labels
+        annotations          = node.annotations
+
+        consolidation = {
+          enabled = node.ttl_seconds_after_empty == null ? true : false
```
Comment (SweetOps) [inline modules/karpenter-provisioners/main.tf:11]: convert `consolidation` to variable and if it's set to true then set `ttlSecondsAfterEmpty` to null

File: modules/karpenter-provisioners/main.tf [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+          blockDeviceMappings = temp.block_device_mappings
+        }
+      }
+    ]
+    }
+  )
+}
+
+module "karpenter_provisioner_label" {
```
Comment (SweetOps) [inline modules/karpenter-provisioners/main.tf:72]: rather doing that you can copy [context](https://github.com/rallyware/terraform-argocd-aws-eks-cluster-bootstrap/blob/main/context.tf) to the this module and iterate over the context, for example:
```hcl
module "node_template_label" {
for_each = { for node in var.karpenter_node_pools : node.name => node }

 source  = "cloudposse/label/null"
 version = "0.25.0"

context = module.this.context
attributes = [each.value.name]
tags = merge(module.this.tags, each.value.kubernetes_labels)


}

File: modules/karpenter-provisioners/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+output "node_templates_yaml" {
+  value       = yamlencode(var.node_templates)
+  description = "YAML encoded Karpenter node templates"
+}
+
+output "node_templates" {
+  value       = var.node_templates
+  description = "Karpenter node templates"
+}
```
Comment (SweetOps) [inline modules/karpenter-provisioners/outputs.tf:24]: missed new line

File: modules/karpenter-provisioners/variables.tf [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+        }))
+        evictionMaxPodGracePeriod   = number
+        imageGCHighThresholdPercent = number
+        imageGCLowThresholdPercent  = number
+        cpuCFSQuota                 = bool
+        podsPerCore                 = number
+        maxPods                     = number
+        })
+        ), {
+        clusterDNS       = ["10.0.1.100"]
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:120]:
```suggestion
        clusterDNS       = null
```

File: modules/karpenter-provisioners/variables.tf [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+        evictionSoftGracePeriod = map(object({
+          "memory.available"  = "5m0s"
+          "nodefs.available"  = "1m30s"
+          "nodefs.inodesFree" = "2m"
+        }))
+        evictionMaxPodGracePeriod   = 600
+        imageGCHighThresholdPercent = 85
+        imageGCLowThresholdPercent  = 80
+        cpuCFSQuota                 = true
+        podsPerCore                 = 2
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:151]:
```suggestion
        podsPerCore                 = null
```

**Review by SweetOps — CHANGES_REQUESTED** [review]

File: modules/karpenter-provisioners/main.tf (hunk 7 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/main.tf:72]: Updated

File: modules/karpenter-provisioners/variables.tf (hunk 3 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf:38]: Updated

File: modules/karpenter-provisioners/variables.tf (hunk 1 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: Updated

File: modules/karpenter-provisioners/variables.tf (hunk 2 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: Updated

File: modules/karpenter-provisioners/variables.tf (hunk 4 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: Updated

File: modules/karpenter-provisioners/variables.tf (hunk 5 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: Updated

File: modules/karpenter-provisioners/main.tf (hunk 6 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/main.tf]: Updated

File: modules/karpenter-provisioners/outputs.tf (hunk 8 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/outputs.tf]: Updated

File: modules/karpenter-provisioners/variables.tf (hunk 10 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: Updated

File: modules/karpenter-provisioners/outputs.tf [hunk 11]
```diff
@@ -0,0 +1,24 @@
+output "karpenter_provisioner_values" {
```
Comment (SweetOps) [inline modules/karpenter-provisioners/outputs.tf:1]:
```suggestion
output "helm_values" {
```

Comment (SweetOps) [reply modules/karpenter-provisioners/outputs.tf:1]: and delete the rest of outputs

File: modules/karpenter-provisioners/variables.tf [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        image_gc_high_threshold_percent = 85
+        image_gc_low_threshold_percent  = 80
+        cpu_cfs_quota                   = true
+        pods_per_core                   = null
+        max_pods                        = 110
+    })
+
+  }))
+  default     = []
+  description = "A list of karpenter node-pool sets"
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:124]:
```suggestion
  description = "A list of karpenter provisioners"
```

File: modules/karpenter-provisioners/variables.tf [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+        ebs = {
+          volumeType          = "gp3"
+          volumeSize          = "35Gi"
+          deleteOnTermination = true
+          encrypted           = true
+        }
+    }])
+  }))
+  default     = []
+  description = "A list of karpenter node templates"
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:173]:
```suggestion
  description = "A list of Karpenter node templates"
```

File: modules/karpenter-provisioners/variables.tf [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
```
Comment (SweetOps) [inline modules/karpenter-provisioners/variables.tf:174]:
```suggestion
```

File: modules/karpenter-provisioners/main.tf [hunk 15]
```diff
@@ -0,0 +1,106 @@
+locals {
+  karpenter_provisioner_values = yamlencode({ provisioners = [for node in var.karpenter_node_pools :
+    {
+      name = node.name
+      spec = {
+        ttlSecondsAfterEmpty = node.consolidation == true ? null : node.ttl_seconds_after_empty
```
Comment (SweetOps) [inline modules/karpenter-provisioners/main.tf:6]:
```suggestion
        ttlSecondsAfterEmpty = node.consolidation ? null : node.ttl_seconds_after_empty
```

**Review by SweetOps — CHANGES_REQUESTED** [review]

File: modules/karpenter-provisioners/variables.tf [hunk 16]
```diff
@@ -0,0 +1,175 @@
+variable "karpenter_node_pools" {
+  type = list(object({
+    name                      = string
+    instance_types            = list(string)
```
[the remark this answers is not in the document]
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf:4]: https://github.com/rallyware/internal-infra/blob/main/components/terraform/apps/variables.tf#L68 - we actually use this key.

File: modules/karpenter-provisioners/variables.tf [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+    kubernetes_labels         = map(string)
```
[the remark this answers is not in the document]
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: updated

File: modules/karpenter-provisioners/main.tf [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+
+          blockDeviceMappings = temp.block_device_mappings
+        }
+      }
+    ]
+    }
+  )
+}
+
+module "karpenter_provisioner_label" {
```
[the remark this answers is not in the document]
Comment (yhrunyk) [reply modules/karpenter-provisioners/main.tf]: updated

File: modules/karpenter-provisioners/outputs.tf (hunk 11 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/outputs.tf:1]: updated

File: modules/karpenter-provisioners/main.tf [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+        ttlSecondsUntilExpired = node.ttl_seconds_until_expired
+
+        ttlSecondsAfterEmpty = node.ttl_seconds_after_empty
+
+        weight = node.weight
+      }
+    }
+    ]
+
+    nodetemplates = [for temp in var.node_templates :
```
[the remark this answers is not in the document]
Comment (yhrunyk) [reply modules/karpenter-provisioners/main.tf]: updated

File: modules/karpenter-provisioners/variables.tf (hunk 13 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: updated

File: modules/karpenter-provisioners/variables.tf (hunk 14 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: updated

File: modules/karpenter-provisioners/main.tf (hunk 15 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/main.tf]: updated

File: modules/karpenter-provisioners/main.tf [hunk 20]
```diff
@@ -0,0 +1,106 @@
+locals {
+  karpenter_provisioner_values = yamlencode({ provisioners = [for node in var.karpenter_node_pools :
```
[the remark this answers is not in the document]
Comment (yhrunyk) [reply modules/karpenter-provisioners/main.tf]: updated

File: modules/karpenter-provisioners/variables.tf [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+variable "node_templates" {
+  type = list(object({
+    name                    = string
+    subnet_selector         = optional(map(any), null)
+    security_group_selector = optional(map(any), null)
+    ami_selector            = optional(map(any), null)
+    ami_family              = optional(string, "AL2")
+    detailed_monitoring     = optional(bool, true)
```
[the remark this answers is not in the document]
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: updated

File: modules/karpenter-provisioners/variables.tf (hunk 12 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: updated

File: modules/karpenter-provisioners/outputs.tf [hunk 22]
```diff
@@ -0,0 +1,4 @@
...
```
Comment (SweetOps) [inline modules/karpenter-provisioners/outputs.tf:1]:
```suggestion
output "helm_values" {
```

**Review by SweetOps — CHANGES_REQUESTED** [review]

File: modules/karpenter-provisioners/variables.tf (hunk 16 above)
[the remark this answers is not in the document]
Comment (yhrunyk) [reply modules/karpenter-provisioners/variables.tf]: you're right, removed

File: modules/karpenter-provisioners/outputs.tf (hunk 22 above)
Comment (yhrunyk) [reply modules/karpenter-provisioners/outputs.tf]: updated

**Review by SweetOps — APPROVED** [review]

**Pull request merged**